File: 16-3sum-closest/16-3sum-closest.py
class Solution:
    def threeSumClosest(self, num: List[int], target: int) -> int:
        '''- sort the array
           - fix a element and optimize our search for other two elements which is closest to target'''
        num.sort()  #sorting the array
        result = num[0] + num[1] + num[2] #predefine res for reference
        for i in range(len(num) - 2):
            j, k = i+1, len(num) - 1
            while j < k:
                sum = num[i] + num[j] + num[k]
                if sum == target:
                    return sum
                
                if abs(sum - target) < abs(result - target):
                    result = sum
                
                if sum < target:#here sum is less than target means we have to increase sum value
                    j += 1
                elif sum > target:
                    k -= 1
            
        return result

File: 16-3sum-closest/16-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n=nums.size();
        sort(nums.begin(),nums.end());
        int result=nums[0]+nums[1]+nums[2];
        
        for (int i=0;i<n-2;i++){
            int j=i+1;
            int k=n-1;
            
            while (j<k){
                int sum= nums[i]+nums[j]+nums[k];
                if (sum==target) return sum;
                
                if (abs(sum-target)<abs(result-target))
                    result=sum;
                
                if (sum<target)
                    j+=1;
                else
                    k-=1;
                
            }
        }
        return result;
    }
};